### ISSUE TITLE
Integers detection can be improved

### ISSUE THREAD

**Issue by codrin-iftimie** [eemeli/yaml#487]
**Describe the bug**
Following yaml 
```yaml
defaultMode: 0777
```
gets parsed as 
```js
{defaultMode: 777}
```

**To Reproduce**
Use `parseDocument` with the default schema.

**Expected behaviour**
I'm not sure from the spec if this is considered an integer, but I have seen ppl using `[+-]?[1-9][0-9]` as the regex for detection.
Ideally this should not be parsed as integer (I assume `parseInt` is used `parseInt('0777', 10) === 777`)
```js
{defaultMode: "0777"}
```

**Versions (please complete the following information):**

- Environment: [Chrome  115.0.5790.102]
Tested two versions:
- `yaml`: "2.1.1"
- `yaml`: "2.3.1"

**Additional context**
This sample YAML is part of a kubernetes manifest where the defaultMode refers to the octal value passed to what I think is a `chmod` command.

**Comment by eemeli** [comment]
This library defaults to the YAML 1.2 core spec, which is rather clear about number parsing: https://yaml.org/spec/1.2.2/#1032-tag-resolution.

YAML 1.1 parsed leading-zero numbers as octal:
```js
YAML.parse('0777', { version: '1.1' }) === 511
```

**Comment by codrin-iftimie** [comment]
Sorry, I phrased that the detection is wrong. Detection is good. 
The parsing is where I expect something different, or doing parse then stringify. I expect that the string does not change. But I think there is a distinction between
```js
YAML.stringify(YAML.parse('0777', { version: '1.1' })) // 511
```
and
```js
YAML.stringify(YAML.parse('"0777"', { version: '1.1' })) // "0777"
```
In the first instance, I'd expect to have `0777` which is still valid in JS number in octal form.

Digging a bit further I learned that `JSON.parse('{"a": 0777}')` will throw.
While 0777 and 0o777 are numbers in javascript (at least I can type them in the console without any issues), there is no function to return them in this format. 
```js
const obj = {a: 0777}
JSON.stringify(obj) // '{"a":511}'
```
